### ISSUE TITLE
Play! Doesn't work on iOS 9 any more

### ISSUE THREAD

**Issue by Fadi5555** [jpd002/Play-#337]
@jpd002 well I updated my iPad Air 2 to iOS 9.0.2 and jailbreaking. So after that when I select Play! To go in its crashing immediately. Any solution please?

**Comment by jpd002** [comment]
I'm currently looking into jailbreaking my device that's on 9.0.2 and I will do some tests. But I've been able to use the emulator while debugging with Xcode though.

**Comment by Fadi5555** [comment]
Ok.
I'll waiting for the news after you test it.

Also did you receive (haunting Ground) or not yet.
Thannks.

**Comment by jpd002** [comment]
I haven't received Haunting Ground yet.

**Comment by Fadi5555** [comment]
@jpd002 the creator of (controller for all) gonna update his tweak to iOS 9
This is picture of this tweak
![image](https://cloud.githubusercontent.com/assets/10546081/10722309/cc5d48b6-7bbf-11e5-98c2-c77bd4b97b70.jpeg)

For more info this is his Twitter: https://twitter.com/cjori

I mean here can you fix issue of the buttons not working?
Because @LoungeKatt is already tried to support the controller through that this tweak (controller for all)
But never work however it's connected successfully.

**Comment by jpd002** [comment]
@Fadi5555 Nice! Still don't know why the emulator crashes as soon as it's launched on a JB 9.0.2 device though. Other emulators seem to have the same problem (namely PPSSPP).

**Comment by Fadi5555** [comment]
@jpd002 I thing because of JIT
To be honest jailbreak by (pangu) always broke the JIT. Unlike (Taig) is so friendly jailbreak and not broke JIT and stable as well.
I always go to ppsspp and retroarch forms and they have same issue of crash so they want to fix JIT to could launch it without any crash.

Anyway here is a link for ppsspp about having same problem.
https://github.com/hrydgard/ppsspp/issues/8066

**Comment by angelXwind** [comment]
@jpd002 Hi, I'm an iOS developer/maintainer/person for PPSSPP (and reicast) — Pangu9 (`fuxiqin`) doesn't patch the kernel anymore, so `VM_PROTECT` is untouched.

Your JIT code will need to be modified to some degree.

I have absolutely no idea how Play! works — literally only just heard of this project today.

I'll finish writing up a new issue entry for PPSSPP (because the current one has gotten kinda messy/off-topic/incorrect) — the info that I will write there applies to you too (and RetroArch, and any other project that wants to do JIT).

**Comment by jpd002** [comment]
@angelXwind Hi! Thanks a lot for your help. I was closely watching your comments on PPSSPP's issue to see what you were going to propose to fix their JIT.

This is basically what I do to enable executable code on iOS and OSX: https://github.com/jpd002/Play--CodeGen/blob/master/src/MemoryFunction.cpp#L49. I'm not familiar with PPSSPP's code, so I don't know if it's the same method, but I'm guessing it's quite similar and that your changes to PPSSPP will be relevant to what I'm doing here.

**Comment by unknownbrackets** [comment]
Right.  My understanding is that `vm_protect()` can't work, so you have to deallocate and then reallocate by pegging the same pointer with the new protection flags.  Then you cross your fingers and hope that no other process changed that memory while you weren't looking.

Currently [we just `mprotect()` everywhere](https://github.com/hrydgard/ppsspp/blob/master/Common/MemoryUtil.cpp), except on Windows, where we `VirtualProtect()`.  Allocation is a bit more complicated because iirc mmap will succeed even when the memory is already allocated by someone else, in some cases.

-[Unknown]

**Comment by Fadi5555** [comment]
That's really useful information. Well I think the problem is from pangu9 jailbreak itself. Because Taig jailbreak always come without break JIT or anything else.

We really hope to find a way to patch VM_PROTECT again to could use JIT. Because this is very important and responsibility for performance of emulators games.

**Comment by Fadi5555** [comment]
@jpd002 did you fix the crash of play! on iOS 9?
If not do you have plan to fix it? Because @angelXwind is give you a solution for this issue.

**Comment by jpd002** [comment]
Not yet, but it's at the top of my TODO list.

**Comment by Fadi5555** [comment]
Ok i hope you'll find a solution for crash issue.
Other thing. Did you fix cover of games not shown?

**Comment by jpd002** [comment]
Not yet. The virtual pad is in now and I think input using controllers has been fixed (I still need to test that though).

**Comment by LoungeKatt** [comment]
Physical controllers were more or less just waiting on the virtual pad. It was never broken, just not actually sending the input on to the emulator.

**Comment by Fadi5555** [comment]
@LoungeKatt got the point.
You did well on supporting the controller but it didn't respond because of the input isn't available yet.
So if the input available the controller will work like a charm.
I was think it's broken but that's really nice to here from you

Well, it's long time no see. welcome back.

**Comment by jpd002** [comment]
Ok, I changed the code-signing identity as per @angelXwind's recommendation and that fixed the "crash at start" issue. 

I then gave the emulator a try on my jailbroken iPhone 5s on iOS 9.0.2 and everything seems to work normally without any modification to the page permission changing code. I've got no idea why it's working though. Feel free to give it a try and see if it works on your devices: http://cydia.purei.org/.

**Comment by unknownbrackets** [comment]
Hmm.  We use mprotect not vm_protect.  Maybe that's why?  This sounds interesting.  But then you just reallocate every time you change jit?  Do you individually allocate each block?  That sounds expensive.

We just use one large buffer, and manage block allocation within it.  We unlock write when jitting new blocks, and then relock/exec it once we're done.  There's almost no copying, as we emit directly to the RAM that will be executable.  Phones, right?  Gotta be fast.

But maybe that's why we're crashing.  He hardcode references between blocks (e.g. for static jumps) so copying would be a big change and perf hit for us.

-[Unknown]

**Comment by jpd002** [comment]
@unknownbrackets The iPhone 5s is 64-bit, but honestly, this needs to be tested by other people, because it doesn't make much sense for my code to work without changing anything, so, my test might be wrong.

**Comment by Fadi5555** [comment]
@jpd002 wow you did a great job.
First of all the emu doesn't crash now anymore.
The dpad buttons now appear on me
Also the controller now I can use some buttons on it but not correctly like when I select up it goes down and cross button it selects square and something like that. I thing it's easy to remap the buttons for the controller.

Well until now the image of cover games still not appearing for games. Any idea please?

**Comment by jpd002** [comment]
@Fadi5555 On which iOS device did you try the emulator on? Which game did you try?

**Comment by unknownbrackets** [comment]
The iPad Air 2 is definitely 64 bit, and even A8X, so it sounds like things are working indeed.

-[Unknown]

**Comment by jpd002** [comment]
@unknownbrackets Yeah, Play! allocates a piece of memory using vm_allocate/VirtualAlloc/mmap for every single "basic block" out there and there's no concept of "patching" a block after it's been compiled.

**Comment by unknownbrackets** [comment]
You're missing out.  Try PPSSPP and switch off "enableBlocklink = true;".  Turning it off gives me a 16% performance hit in Fat Princess, for example.  That's just patching static jumps between basic blocks to not go through the dispatcher.

-[Unknown]

**Comment by Suhayl** [comment]
It crashes on my iPhone 6. :(

**Comment by Fadi5555** [comment]
![image](https://cloud.githubusercontent.com/assets/10546081/10901935/fe7b793c-8207-11e5-8bdb-198b96d04f17.png)
![image](https://cloud.githubusercontent.com/assets/10546081/10901938/0d823fd8-8208-11e5-8202-f7901a683c13.png)

**Comment by Suhayl** [comment]
Ooh cool! How did you get it to work?

**Comment by Fadi5555** [comment]
@Suhay! use @jpd002 repo http://cydia.purei.org/ and download Play! That's it.

**Comment by Suhayl** [comment]
I did. I'm getting the white screen that says play and everything and it just crashes on me. Maybe I didn't respring or something?

**Comment by Fadi5555** [comment]
Hmm. Try to download powerapp from cydia then select refresh cache then respring

**Comment by Suhayl** [comment]
Didn't do that yet. I'll do it now.

**Comment by Suhayl** [comment]
Tried it. Still nothing. I wish there was a way I could record my screen and show you guys what's going on.

**Comment by Fadi5555** [comment]
Hmmm which version of your iPhone?

**Comment by Suhayl** [comment]
It's an iPhone 6 and on iOS 9.0.2
Lol and obviously jailbroken.

**Comment by Fadi5555** [comment]
Can you make screenshot for it

**Comment by Suhayl** [comment]
This is what I get. And then it crashes back to the springboard.

**Comment by Fadi5555** [comment]
Well you can ask @jpd002 for help
I tried to help
Sorry

**Comment by Suhayl** [comment]
Hey don't be sorry. I appreciate the help. Maybe in a weekly build it'll be fixed or maybe I'm just being dumb and doing something wrong.

**Comment by scorpio16v** [comment]
Play! 0.30-20151102-signtest1 works now. Tested with iPod Touch 6.Gen on iOS 9.0.2. and Gradius V .iso .

**Comment by xXPaladinXx** [comment]
Dragon Ball Z infinite world is playable of course slow but still works. I'm still having the issue where some of my games won't appear am I the only one?

**Comment by Suhayl** [comment]
What phone do you have?

**Comment by xXPaladinXx** [comment]
I have an iPod 6 running iOS 9.0.2

**Comment by Suhayl** [comment]
Ohh.

**Comment by xXPaladinXx** [comment]
Is their like no fix for this?

**Comment by Suhayl** [comment]
There probably is. But I'm still trying to get Play! to work on my iPhone 6. It keeps crashing back to the springboard

**Comment by Suhayl** [comment]
@jpd002 I kinda need help with something

**Comment by jpd002** [comment]
@Suhayl Did you get my e-mail response? Basically, I was asking if you could give me a crash log so I can get a better idea of what the crash could be. Thanks!

**Comment by Suhayl** [comment]
Oh no I didn't get your email response I'm sorry. Give me a sec. I dunno if there is a crash log

**Comment by Suhayl** [comment]
Try my email: suhayl.khatib@com

**Comment by Fadi5555** [comment]
@jpd002 I'm exited to see the fix of cover image of games
@Suhayl I hope you can find a solution for your own issue by jpd002 just give himself the crash log to know what is kind of crash you got.

**Comment by Suhayl** [comment]
@Fadi5555 dumb question but where do I find the crash log? 😅️

**Comment by Suhayl** [comment]
Also, fun fact: I have crashreporter on my iPhone and not once has it said Play! crashed.

**Comment by jpd002** [comment]
@Suhayl Humm, in that case, would you happen to have a excerpt of the device's log at the moment you're trying to launch the application?

**Comment by Suhayl** [comment]
I could try. Where do I find it?

**Comment by jpd002** [comment]
I usually use Xcode to do that, but if you don't have it, maybe that can help: https://www.theiphonewiki.com/wiki/System_Log#Option_A:_from_your_device

**Comment by Suhayl** [comment]
I think I got it. What's the email to send it to?

**Comment by jpd002** [comment]
@Suhayl Same one you used to send me the Cydia logs.

**Comment by Fadi5555** [comment]
@jpd002 after update play! To latest one. The images of games still missing.
I recommend to give me screenshot of your file of game through (ifile) also give me the shot of the permission of them and shot for your play! Screen to see the images of games of yours.
That's only to make sure what happened?

**Comment by jpd002** [comment]
@Fadi5555 Humm, that's odd. The covers are missing for all of the games?

**Comment by Fadi5555** [comment]
@jpd002 just look: this is the last update form you.
This is from my iPad Air 2 on iOS 9.0.2
 
![image](https://cloud.githubusercontent.com/assets/10546081/11056545/0d055a54-8794-11e5-99ea-f28de969498a.png)

And this from my iPhone 5s on iOS 8.4

![image](https://cloud.githubusercontent.com/assets/10546081/11056616/cd229dd8-8794-11e5-9d6f-9f74dd8568a9.png)

Any solution please?

**Comment by Fadi5555** [comment]
@jpd002 isn't there any new things on your project?
We really miss your progress on this work
Please come back.

**Comment by DPersonalized** [comment]
@Fadi5555 Please don't add to this unless you have something useful to say.

**Comment by Fadi5555** [comment]
@DPersonalized really? What do you think is useful to say?
I know what I wrote so please don't say something like that to me again

**Comment by JohnPaulHarold** [comment]
@Fadi5555 The purpose of a GitHub issues tracker is to highlight issues with the code, not to chime in with requests for general updates. You're better off finding a forum for that, or reddit, or something. It'd be nice if you kept the forum style posts off GitHub, and only raise issues pertaining to code quality, implementations, or bugs in released binaries. Hope that helps.

**Comment by jpd002** [comment]
I've been working on improving compatibility recently, but I haven't pushed many commits because I need to do some testing before. Also been playing with Vulkan this past week but it's still in very early stages. Sound on Android, iOS and OSX should be coming soon too.

**Comment by Fadi5555** [comment]
@jpd002 welcome back.
It's very nice to hear that from you indeed.

On iOS it needs to mapping some buttons I left some comments some where about it
And Vulkan is similar as metal on iOS I mean maybe we can get more benefit if we support it.
Finally the supporting sound looks fantastic. Keep going.

**Comment by jpd002** [comment]
@Suhayl Have you tried the latest weekly build (2016-05-08)? I've made changes to the way files are scanned and it might fix the crash you were having.
